Keyword priority (#1130)
neo4j/docs-cypher: modules/ROOT/pages/syntax/keywords.adoc

:description: This section contains a list of reserved keywords in Cypher.

[[cypher-keywords]]
= Keywords

> Keywords are words that have a special meaning in Cypher.

Reviewer: ```suggestion
Keywords are words with a predefined meaning in Cypher.
```

> The keywords are not recommended to be used as identifiers in the following contexts:

Reviewer: ```suggestion
Keywords are not recommended to be used as identifiers in the following contexts:
```

> The keywords are not recommended to be used as identifiers in the following contexts:

Reviewer: is the recommendation strong enough here? if it doesn't work at all, we could say that here and then offer the use of quotes below as a way of still using the keyword as an identifier, if the user wants to do that (if so, remove "unquoted" from line 14)

Author: How do you mean strong enough? I am not sure I follow.

After all it is a recommendation. User do not have to do that but it will make their live a little less complicated. The only true safe thing are quote identifiers. But Valerio seems not to be in favour of recommending that. Hence, I only hint at it and leave to the reader make this last step of logical inference.

Reviewer: in the contexts pointed out, it only works when using quotations because otherwise they would be interpreted as the keywords, correct?

in a sense, it's not a recommendation then, rather it just doesn't work without quotations

Author: Yes. In the specific contexts of ambiguity, which are more subtle then what is listed here, you have to use quotation to get identifier interpretation. 

However, the recommendation is deliberately more coarse grained, recommend to us quotes on identifiers (that are labels, relationship types, and variable) equal to keywords in all places.

For instance, the commendation is to not do something like this:
```
WITH 1 AS RETURN
RETURN RETURN
```
although is works perfectly fine.

As the example hopefully illustrates, the recommendation is more coarse grained than functionally needed, because 
1. it can be confusing for a query reader if you use keywords as identifiers without quotes even if there is no ambiguity for the system
2. you do not have to worry about in where and what these exact contexts of actual ambiguity are, i.e. the recommendation is much simpler and more stable of time

BTW: 
I removed functions names from the recommendation, but maybe we should add that back in, because the following is likewise confusing:
```
WITH WITH("foo") AS RETURN
CALL CALL("bar") YIELD YIELD
RETURN RETURN, YIELD
```
My thinking was that most users do not assign function names. Only if you do have custom functions (UDF) you assigned function names. Nevertheless. I will add function name and procedure them back in.

Author: It also included parameters, originally. These can never be ambiguous because they are prefixed with a `$`.

So do we think that 
```
WITH $WITH AS foo
RETURN $RETURN, $AS, foo
```

is to be avoided? I am open to opinions here.

Note that these are deliberately arcane examples.

There are many instance that are less arcane and confusing, e.g.:
```
WITH $USER AS foo
RETURN $TEXT, foo
```
or (without parameters)
```
WITH "bob" AS USER, "Hello" AS START
RETURN START + USER AS SHOW
```

I guess, we do not want to scare user too much about the need of quoting. But that want something to point to if they start to complain about too weird cases — which most users will avoid their own sake.

Reviewer: alright, let's leave it as is then :) thanks for taking the time to elaborate

* Variables
* Labels
* Relationship types

> If a snippet of Cypher can validly be a keyword as well as an unquoted identifier, it is always interpreted as a keyword.

Reviewer: ```suggestion
If a snippet of Cypher can both be a keyword and an unquoted identifier, it is interpreted as a keyword.
```

> If a snippet of Cypher can validly be a keyword as well as an unquoted identifier, it is always interpreted as a keyword.

Reviewer: wondering about "snippet" - the term suggests that it can be more than one word, while "keyword" suggests that it's a single one (or rather a string which doesn't contain spaces, thinking of "ALL_SHORTEST_PATHS" etc)

would "expression" work?

Author: "expression" has a precise meaning. `a + b` is an expression to, while e.g. a label identifier alone is not an expression. We do not have a proper term, so I picked something non-technical. 

However, I am not married to "snippet". Not really a fan of my choice here in fact.

- A piece of Cypher 
- A part of Cypher
- Cypher text
- ...

any would work. 

Or a totally difference angle may be this:

> If an unquoted identifier equal to a keyword is user in Cypher in a place where can also the keyword, then it is interpreted as the keyword.

BTW: Feel free to edit the PR directly.

Reviewer: ah, yes, "expression" wouldn't work, you're right :/

what about "piece of Cypher syntax"?

For instance in the following, `true` could be the variable `true` or the Boolean literal value `true`. Since `true` is a keyword, the query returns a single Boolean `true` value.

.Query
[source, cypher, role=test-result-skip]
----
WITH 123 AS true

> RETURN true AS x LIMIT 1

Reviewer 2: Do the LIMIT have some kind of effect on this example?

Author: LIMIT got removed.

----

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===
| x

| true

1+d|Rows: 1
|===

> If any keyword is quoted in backticks (```), such as `++`true`++`, it is always interpreted as an identifier in the given context.

Reviewer: ```suggestion
If a keyword is quoted in backticks (```), such as `++`true`++`, it is interpreted as an identifier in the given context.
```

Author: I am fine with your suggestion. 

However, if we like my proposal above, we may want to do something like following here:

> A quoted identifier equal to a keyword is interpreted as an identifier (even if it is equal to a keyword).

Reviewer: alright, let's leave it as is then :) thanks for taking the time to elaborate

.Query
[source, cypher, role=test-result-skip]
----
WITH 123 AS true
RETURN `true` AS x LIMIT 1
----

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===
| x

| 123

1+d|Rows: 1
|===

== List of keywords

> The following lists of all keywords in Cypher.

Reviewer 2: It would be nice with a test somewhere that fails when a new keyword is added and remind us to update this list.

Author: I think, I can come up with something like that. I will create a card.

> The following lists of all keywords in Cypher.

Reviewer: ```suggestion
The following lists all keywords in Cypher.
```

> Note that with future functionality, Cypher may gain additional keywords not yet in the list.

Reviewer: ```suggestion
Note that with future functionality, Cypher may be extended with additional keywords.
```
